Build/Versions.cs

```diff
@@ -0,0 +1,77 @@
+﻿using Semver;
+using static SimpleExec.Command;
+
+namespace Build;
+
+public static class Versions
+{
+  private static string? s_currentTag;
+  private static SemVersion? s_currentVersion;
+
+  public static async Task<string> GetCurrentTag()
+  {
+    if (s_currentTag is not null)
+    {
+      return s_currentTag;
+    }
+    var (currentTag, _) = await ReadAsync("git", "describe --tags");
+    currentTag = currentTag.Trim();
+    s_currentTag = currentTag;
+    return s_currentTag;
+  }
+
+  public static async Task<SemVersion> ComputeVersion()
+  {
+    if (s_currentVersion is not null)
+    {
+      return s_currentVersion;
+    }
+    var currentTag = await GetCurrentTag();
+
+    if (!SemVersion.TryParse(currentTag, SemVersionStyles.AllowLowerV, out var currentVersion))
+    {
+      throw new InvalidOperationException($"Could not parse version: '{currentTag}'");
+    }
+    s_currentVersion = currentVersion;
+    return s_currentVersion;
+  }
+
+  private static string? s_currentFileVersion;
+
+  public static async Task<string> ComputeFileVersion()
+  {
+    if (s_currentFileVersion is not null)
+    {
+      return s_currentFileVersion;
+    }
+    var currentVersion = await ComputeVersion();
+    s_currentFileVersion = currentVersion.WithoutPrereleaseOrMetadata() + ".0";
+    return s_currentFileVersion;
+  }
+
+  public static async Task<string> GetPreviousTag(string currentTag)
+  {
+    var (lastTag, _) = await ReadAsync("git", $"describe --abbrev=0 --tags {currentTag}^");
```

Comment:
does this consider the versioned tags? as use case scenario, if someone push a tag for any reason to anchor a commit, we could get the wrong state by any chance?

Reply:
a tag is a tag....we parse it and it fails the build